aten/src/ATen/native/native_functions.yaml

```diff
@@ -175,6 +175,14 @@
     CPU: _assert_async_msg_cpu
     CUDA: _assert_async_msg_cuda
 
+- func: _assert_scalar(Scalar self, str assert_msg) -> ()
+  dispatch:
+    CompositeExplicitAutograd: _assert_scalar
+
+- func: _functional_assert_scalar(Scalar self, str assert_msg, Tensor dep_token) -> Tensor
```

Comment:
fwiw, one bit about the "functionalized asserts" design that still felt/feels unclear to me is that:

(a) it uses a global dependency token to promise we have a global ordering across assertions, that a compiler can't reorder

(b) but this still does **not** prevent the compiler from reordering an assert w.r.t. other non-assert-related nodes around it (which seems bad, since if we have an assert that x > 0 before using x, we don't want to move that assert to be later).

Reply:
That's right. IIRC, we discussed alternate schemes; e.g., if you want to assert on Tensor data, you should have something like `_assert_scalar_tensor(Scalar cond, str assert_msg, Tensor validated) -> Tensor` and you are supposed to use the output of this operation, which guarantees subsequent reads from the output are post the assert in question.

The other alternative is to teach Inductor not to reorder across asserts (implicit barriers).

I'm not sure we decided one way or another (cc @Chillee @jansel)